### PR TITLE
Remove canDelete property so that non-summary pages can always be deleted.

### LINKED ISSUE #147: Delete page sometimes only hides it
afbeelding

The delete page button above does not work for some pages, but does work for others. Need to dig into this to find out why.

### PR DESCRIPTION
**Pull request by Dorus**
Fix #147

### CODE REVIEW

File: Yafc/Windows/ProjectPageSettingsPanel.cs [hunk 1]
```diff
@@ -71,7 +71,7 @@ public class ProjectPageSettingsPanel : PseudoScreen {
 
                 gui.allocator = RectAllocator.LeftRow;
                 if (editingPage != null && gui.BuildRedButton("Delete page")) {
-                    if (editingPage.canDelete) {
+                    if (editingPage.contentType != typeof(Summary)) {
```
Comment (veger) [inline Yafc/Windows/ProjectPageSettingsPanel.cs:74]: So back then I wanted to prevent to adding specific support in a generic location. If no-one objects this is fine by me...

Comment (have-fun-was-taken) [reply Yafc/Windows/ProjectPageSettingsPanel.cs:76]: For future reference, why is specific support in a generic location considered bad? Genuinely curious.

Comment (veger) [reply Yafc/Windows/ProjectPageSettingsPanel.cs:76]: It adds dependencies/couple between unrelated pieces of code. Now some settings panel needs to know about the Summary page. If the Summary page is changed, we might need to change this as well.

Before, a property was introduced to add a generic feature (`canDelete`) and this code only needed to support this feature, without know about the page(s) that require it.

Comment (Dorus) [reply Yafc/Windows/ProjectPageSettingsPanel.cs:76]: I guess a more proper way to do it is to put the canDelete back in the page, but change the check to what's writen here. That way settingPanel stays decoupled from editingPage, and the editingPage only needs to know about deletion, not the how. I can change that.

**Comment by Dorus** [comment]
afbeelding
Now hitting this code on my test project. Not sure if it's related. Will need to check.

**Comment by Dorus** [comment]
I've reduced the above mentioned problem to 3 lines of

```
            {
              "multiplier": 1,
              "page": null,
              "subgroup": {
                "elements": [],
                "expanded": false,
                "name": null
              }
            },
            {
              "multiplier": 1,
              "page": null,
              "subgroup": {
                "elements": [],
                "expanded": false,
                "name": null
              }
            },
            {
              "multiplier": 1,
              "page": null,
              "subgroup": {
                "elements": [],
                "expanded": false,
                "name": null
              }
```
In one of the `YAFC.Model.ProductionSummary` -> `content.group.elements` items. However saving the file again fixes it and i cannot figure out how i made this happen.

I'm running in some other issues while testing this, will report those separate since they're also on master. Meaning it's unrlated to this PR and this Pr could be merged.

**Comment by have-fun-was-taken** [comment]
> saving the file again fixes it

Then we need something like a one-off patching mechanism? For instance, if it detects that the project file was updated earlier than a certain date, it makes a backup and re-saves the file?

**Comment by veger** [comment]
The 3 elements you shows do not have undefined subgroups? So `subgroup == null` should be `false` and the exception should not have been thrown?

**Review by veger — APPROVED** [review]
I do not like the `contentType != typeof(Summary)` check a lot, but it is the easiest way :smile:

**Comment by Dorus** [comment]
> > saving the file again fixes it
> 
> In addition to the changes in the PR, what do you think about a patching mechanism for the saves? For instance, if YAFC detects that the project file was updated earlier than the date this fix was released, then it makes a backup and re-saves the file.

Even the master branch fixes it on save. So the cause of that problem was either me manipulating the process in the debugger, or introducing it in one of my earlier changes, or it might even  have been something in the older master before i rebased. Either way i cannot reproduce it anymore both on master and this branch.

The only thing i can do, is introduce some robustness in YaFC when it hits the notsuportedexception, when i step past the notsuportedexception in YaFC with the debugger and save again, the problem fixes itself, so that's something i could patch in. But that seems more appropriate for a separate pr. (Sorry i should have mentioned it only fixes it on saving if i manipulate the program in de debugger during loading, else it wont load the file at all)".

Tl;dr: i dont think it's a problem introduced by the current pr. I could fix the _loading_ code, but would do so in a separate pr.

**Comment by have-fun-was-taken** [comment]
Rebased on fresh master
